tainting: Improvements to taint_assume_safe_{booleans,numbers}

IagoAbal opened the pull request:
Mainly:

- Use type info provided by explicit casts.
- Use const-prop info to infer types.
- Guess the type of `length` when used as an attribute.
- Add `long` and `Long` as "synonyms" for `int` and `Integer`, and `double` for `float`.
- Guess types for `decode`, `valueOf` and `parseXyz` methods in classes `Boolean`, `Integer`, `Long`, and `Float`.
- Sometimes we cannot infer the type of the RHS but the LHS is a variable with e.g. a number type, so we can avoid tracking it.

Follows: 64cb387f7a0 ("tainting: Add options to ignore taint based on types (#7936)")

test plan:
make test # new tests

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

nmote commented on src/typing/Typing.ml:
```diff
@@ -22,22 +22,19 @@ module G = AST_generic
-      in
+      let t = type_of_lit lit in
       (t, None)
+  | G.DotAccess
+      (_obj, _, FN (Id (("length", _), { id_type = { contents = None }; _ })))
+    when lang =*= Lang.Java ->
```
As discussed I'd appreciate a comment here explaining how this does not conflict with the existing guessing for e.g. `string.length()`

nmote reviewed (approved):
As we discussed I think this is the best option for the moment. Thanks for talking through the alternatives with me.

IagoAbal committed:
tainting: Improvements to taint_assume_safe_{booleans,numbers}

Mainly:

- Use type info provided by explicit casts.
- Use const-prop info to infer types.
- Guess the type of `length` when used as an attribute.
- Add `long` and `Long` as "synonyms" for `int` and `Integer`,
  and `double` for `float`.
- Guess types for `decode`, `valueOf` and `parseXyz` methods in
  classes `Boolean`, `Integer, `Long`, and `Float`.
- Sometimes we cannot infer the type of the RHS but the LHS is a
  variable with e.g. a number type, so we can avoid tracking it.

Follows: 64cb387f7a0 ("tainting: Add options to ignore taint based on types (#7936)")

test plan:
make test # new tests

IagoAbal committed:
Nat's review
